### ISSUE TITLE
Max_client_disconnect ignores no-reschedule policy.

### ISSUE THREAD

**Issue by stswidwinski** [hashicorp/nomad#16515]
## Advertised contract.

The advertised contract is as follows:

> These allocations may continue to run on the disconnected client. Replacement allocations will be scheduled according to the allocations' reschedule policy until the disconnected client reconnects. Once a disconnected client reconnects, Nomad will compare the "unknown" allocations with their replacements and keep the one with the best node score. If the max_client_disconnect duration expires before the client reconnects, the allocations will be marked "lost". Clients that contain "unknown" allocations will transition to "disconnected" rather than "down" until the last max_client_disconnect duration has expired.

As described here: https://developer.hashicorp.com/nomad/docs/job-specification/group#max-client-disconnect.

However, we have observed that this contract is simply not obeyed and workloads are restarted immediately at the time of node becoming unavailable. This is due to the fact that the logic which handles evaluations for `node-updated` will cause placement of allocations which are *not* related to those which are "lost" (for the disconnected node).

## Root cause.

On node state update (which occurs when the node transitions from ready  to disconnected ):

https://github.com/hashicorp/nomad/blob/v1.5.1/nomad/node_endpoint.go#L479

Nomad will issue an evaluation:

https://github.com/hashicorp/nomad/blob/v1.5.1/nomad/node_endpoint.go#L577-L587

https://github.com/hashicorp/nomad/blob/v1.5.1/nomad/node_endpoint.go#L661-L667

This evaluation will add evaluations per-job as described here:

https://github.com/hashicorp/nomad/blob/v1.5.1/nomad/node_endpoint.go#L1593-L1595

Now, the evaluation computation is where the bug lives. The computation will go through the regular allocation reconciler here:

https://github.com/hashicorp/nomad/blob/v1.5.1/scheduler/generic_sched.go#L385

And eventually to computing the allocations for placement within computeDeploymentComplete  and computeGroup:

https://github.com/hashicorp/nomad/blob/v1.5.1/scheduler/reconcile.go#L434-L436

When we filter for disconnecting nodes we ignore the returned values of untainted . This is the crux of the bug since later on when we compute the placement of allocations we will get the following variables within:

https://github.com/hashicorp/nomad/blob/v1.5.1/scheduler/reconcile.go#L502

```
(dlv) print untainted
github.com/hashicorp/nomad/scheduler.allocSet []
(dlv) print migrate
github.com/hashicorp/nomad/scheduler.allocSet []
(dlv) print reschedule
github.com/hashicorp/nomad/scheduler.allocSet []
(dlv) print lost
github.com/hashicorp/nomad/scheduler.allocSet []
(dlv) print reconnecting
github.com/hashicorp/nomad/scheduler.allocSet []
```

This will lead to the following code adding allocations to ensure that we have at least Group.count  number of allocations:

https://github.com/hashicorp/nomad/blob/v1.5.1/scheduler/reconcile.go#L752-L760

So this is what adds allocations when there should be no allocations added. Humorously, much later we actually also schedule an evaluation for later to handle the disconnecting  state in relationship with the disconnected  clients:

https://github.com/hashicorp/nomad/blob/v1.5.1/scheduler/generic_sched.go#L314-L324

## Repro.

We start up a cluster with 1 server and 2 clients:
```
# Show server config
$ cat ./server.conf
data_dir = "/tmp/nomad/server"

advertise {
  http = "127.0.0.1"
  rpc = "127.0.0.1"
  serf = "127.0.0.1"
}

server {
  enabled = true
  bootstrap_expect = 1
}

# Start server
./nomad agent -config server.conf


# Show first client config
$ cat ./client-1.conf 
data_dir = "/tmp/nomad/client-1"
log_level = "debug"

advertise {
  http = "127.0.0.1"
  rpc = "127.0.0.1"
  serf = "127.0.0.1"
}

ports {
  http = "9876"
  rpc = "9875"
  serf = "9874"
}

client {
  enabled = true
  servers = ["127.0.0.1"]
  gc_max_allocs = 1
}

plugin "raw_exec" {
  config {
    enabled = true
  }
}

# Start first client 
$ ./nomad agent -config client-1.conf


# Show second client config
$ cat ./client-2.conf 
data_dir = "/tmp/nomad/client-2"
log_level = "debug"

advertise {
  http = "127.0.0.1"
  rpc = "127.0.0.1"
  serf = "127.0.0.1"
}

ports {
  http = "8876"
  rpc = "8875"
  serf = "8874"
}

client {
  enabled = true
  servers = ["127.0.0.1"]
  gc_max_allocs = 1
}

plugin "raw_exec" {
  config {
    enabled = true
  }
}

$ Start second client
$ ./nomad agent -config client-2.conf
```

We submit a job with max_client_disconnect  set and the reschedule policy being such that:

1. The delay on restart is huge
2. We should not allow any reschedules in the first place

```
# Show the job spec
$ cat job.hcl 
job "example" {
  datacenters = ["dc1"]
  type = "batch"

  group "test-group" {
    task "test-task" {
      driver = "raw_exec"

      config {
	command = "/usr/bin/sleep"
        args = [ "inf" ]
      }
    }

    reschedule {
      attempts  = 0
      unlimited = false
      delay = "24h"
    }

    max_client_disconnect = "24h"
  }
}

# Submit the job spec 
$ ./nomad job run job.hcl
==> 2023-03-15T22:14:20-04:00: Monitoring evaluation "d8b6f482"
    2023-03-15T22:14:20-04:00: Evaluation triggered by job "example"
    2023-03-15T22:14:21-04:00: Allocation "f06d1eef" created: node "10986d2d", group "test-group"
    2023-03-15T22:14:21-04:00: Evaluation status changed: "pending" -> "complete"
==> 2023-03-15T22:14:21-04:00: Evaluation "d8b6f482" finished with status "complete"

# Inspect the status
$ ./nomad job status example
ID            = example
Name          = example
Submit Date   = 2023-03-15T22:14:20-04:00
Type          = batch
Priority      = 50
Datacenters   = dc1
Namespace     = default
Status        = running
Periodic      = false
Parameterized = false

Summary
Task Group  Queued  Starting  Running  Failed  Complete  Lost  Unknown
test-group  0       0         1        0       0         0     0

Allocations
ID        Node ID   Task Group  Version  Desired  Status   Created  Modified
f06d1eef  10986d2d  test-group  0        run      running  7s ago   7s ago
```

Kill the client which the allocation is running on. It does not matter whether the shutdown is triggered gracefully via SIGINT, SIGTERM or if the shutdown is triggered forcefully via SIGKILL. Then wait for the heartbeat timeouts to expire. Inspect the job and evaluations:

```
# Kill first client...

# Wait for some time for the scheduler to see the changes in state.
#
# NOTE: We get 1 running allocation which is on a new node. We also get 1 allocation that is unknown
$ ./nomad job status example
ID            = example                          
Name          = example                          
Submit Date   = 2023-03-15T22:14:20-04:00       
Type          = batch                                                                                                                              
Priority      = 50                                            
Datacenters   = dc1                                   
Namespace     = default                         
Status        = running                                                 
Periodic      = false                  
Parameterized = false                       
                                                                         
Summary                                                         
Task Group  Queued  Starting  Running  Failed  Complete  Lost  Unknown
test-group  0       0         1        0       0         0     1
                                                                         
Allocations                                         
ID        Node ID   Task Group  Version  Desired  Status   Created    Modified
2554d770  96696c3a  test-group  0        run      running  16s ago    16s ago
f06d1eef  10986d2d  test-group  0        run      unknown  1m45s ago  16s ago
 
# Inspect the evaluations that have happened.
$ ./nomad eval list
ID        Priority  Triggered By            Job ID   Namespace  Node ID   Status    Placement Failures
572f76df  50        max-disconnect-timeout  example  default    <none>    pending   false
1d132b1d  50        node-update             example  default    10986d2d  complete  false
d8b6f482  50        job-register            example  default    <none>    complete  false

$ ./nomad eval status 572f76df
ID                 = 572f76df
Create Time        = 41s ago
Modify Time        = 41s ago
Status             = pending
Status Description = created for delayed disconnect timeout
Type               = batch
TriggeredBy        = max-disconnect-timeout
Job ID             = example
Namespace          = default
Priority           = 50
Placement Failures = false
Wait Until         = 2023-03-16T23:15:08-04:00

$ ./nomad eval status 1d132b1d
ID                 = 1d132b1d
Create Time        = 46s ago
Modify Time        = 46s ago
Status             = complete
Status Description = complete
Type               = batch
TriggeredBy        = node-update
Job ID             = example
Namespace          = default
Node ID            = 10986d2d-29fd-1051-1af1-9d39cd078695
Priority           = 50
Placement Failures = false
 ```

So there are two separate evaluations. One evaluation which correctly waits for max_client_disconnect  and one which ignores it entirely. The latter causes a new allocation to be run defeating the purpose of reschedule stanza in combination with max_client_wait . This also is a direct contradiction with the contract as stated. The logs of Nomad that are relevant:

```
    2023-03-15T22:15:49.712-0400 [DEBUG] worker: dequeued evaluation: worker_id=dfecb445-132d-5d9b-6f4d-9a8fb128a0f7 eval_id=1d132b1d-b46c-f1da-f07e-35fab54be842 type=batch namespace=default job_id=example node_id=10986d2d-29fd-1051-1af1-9d39cd078695 triggered_by=node-update                   
    2023-03-15T22:15:49.713-0400 [DEBUG] worker.batch_sched: reconciled current state with desired state: eval_id=1d132b1d-b46c-f1da-f07e-35fab54be842 job_id=example namespace=default worker_id=dfecb445-132d-5d9b-6f4d-9a8fb128a0f7                                                                
  results=                                                                                                                                                                                                                                                                                            
  | Total changes: (place 1) (destructive 0) (inplace 0) (stop 0) (disconnect 1) (reconnect 0)                                                                                                                                                                                                        
  | Desired Changes for "test-group": (place 1) (inplace 0) (destructive 0) (stop 0) (migrate 0) (ignore 0) (canary 0)                                                                                                                                                                                
                                                                                                                                                                                                                                                                                                      
    2023-03-15T22:15:49.714-0400 [DEBUG] worker: created evaluation: worker_id=dfecb445-132d-5d9b-6f4d-9a8fb128a0f7 eval="<Eval \"572f76df-4777-da86-4980-8415eb6d5f06\" JobID: \"example\" Namespace: \"default\">"                                                                                  
    2023-03-15T22:15:49.714-0400 [DEBUG] worker.batch_sched: codefound reschedulable allocs, followup eval created: eval_id=1d132b1d-b46c-f1da-f07e-35fab54be842 job_id=example namespace=default worker_id=dfecb445-132d-5d9b-6f4d-9a8fb128a0f7 followup_eval_id=572f76df-4777-da86-4980-8415eb6d5f06    
    2023-03-15T22:15:49.715-0400 [DEBUG] worker: submitted plan for evaluation: worker_id=dfecb445-132d-5d9b-6f4d-9a8fb128a0f7 eval_id=1d132b1d-b46c-f1da-f07e-35fab54be842                                                                                                                           
    2023-03-15T22:15:49.715-0400 [DEBUG] worker.batch_sched: setting eval status: eval_id=1d132b1d-b46c-f1da-f07e-35fab54be842 job_id=example namespace=default worker_id=dfecb445-132d-5d9b-6f4d-9a8fb128a0f7 status=complete                                                                        
    2023-03-15T22:15:49.716-0400 [DEBUG] worker: updated evaluation: worker_id=dfecb445-132d-5d9b-6f4d-9a8fb128a0f7 eval="<Eval \"1d132b1d-b46c-f1da-f07e-35fab54be842\" JobID: \"example\" Namespace: \"default\">"                                                                                  
    2023-03-15T22:15:49.716-0400 [DEBUG] worker: ack evaluation: worker_id=dfecb445-132d-5d9b-6f4d-9a8fb128a0f7 eval_id=1d132b1d-b46c-f1da-f07e-35fab54be842 type=batch namespace=default job_id=example node_id=10986d2d-29fd-1051-1af1-9d39cd078695 triggered_by=node-update 
```

**Comment by tgross** [comment]
Hi @stswidwinski!

> So there are two separate evaluations. One evaluation which correctly waits for max_client_disconnect  and one which ignores it entirely.

For a bit of context, the "triggered by" field for the evaluation is for information only. Evaluations are idempotent over the same cluster snapshot, regardless of trigger. In this case, the heartbeat failure only creates an evaluation with [`node-update`](https://github.com/hashicorp/nomad/blob/v1.5.1/nomad/node_endpoint.go#L1652). It's the _scheduler_ that, after evaluating the cluster state, decides (a) whether it needs to create a replacement and (b) write a `max-disconnect-timeout` evaluation that will delay until the timeout expires. So the bug here is not that the `node-update` evaluation is created, because that's how the whole thing gets figured out in the first place.

As you pointed out the `filterByRescheduleable` call over disconnecting allocs (ref [`reconcile.go#L435`](https://github.com/hashicorp/nomad/blob/v1.5.1/scheduler/reconcile.go#L435)) does throw away the "untainted" set:

```go
// Determine what set of disconnecting allocations need to be rescheduled
_, rescheduleDisconnecting, _ := disconnecting.filterByRescheduleable(a.batch, true, a.now, a.evalID, a.deployment)
```

That means that the disconnected alloc we're not replacing isn't getting counted against the total count and so the scheduler is seeing this as if the job was short. In other words, it's not respecting rescheduling because it doesn't see the replacement as a reschedule at all! If you look your allocation `2554d770` in detail, you'll find that it's missing the `PreviousAllocation` field that rescheduled allocations have.

So I think what we're going to need to do here is figure out a way to count the `untainted` returned by `disconnnecting.filterByReschedulable` without messing up anything else in the process. But I think we should be able to safely put it in the ignored set (or destructive update set, if the job has actually changed!) and get the right results. This is going to take a bit of testing to verify.

**Comment by kholisrag** [comment]
Hi @tgross ,

I think we're facing similar issue about this one,
what we want is the batch job (`reschedule attempts = 0` and `unlimited = false`) don't need to schedule the replacement,
but its actually reschedule the replacement as above bug explained.

is my assumption correct ?

**Comment by tgross** [comment]
That's correct. There is no workaround as far as I know (other than not using `max_client_disconnect`, of course).

**Comment by IgnacioHeredia** [comment]
Same issue here.

Steps I executed (on Nomad 1.3):
* Ran a job with no reschedule policy and max_client_disconnect
* Job successfully allocated on `client_1`
* I manually removed the network of `client_1`
* `client_1` is flagged as disconnected (as expected)
* Job is allocated in `client_2` (❗ ignoring the no reschule policy)
* I manually add back the network of `client_1`
* Allocation in `client_1` is successfully recovered, and the previous allocation in `client_2` is discarded

So everything works as expected except for the fact that the job is rescheduled while the client is disconnected, which I was trying to avoid with the no reschedule policy. 

I cannot remove max_client_disconnect because then allocation would be marked as lost, and wouldn't be able to recover it when  `client_1` comes back.

Is there an ETA for this fix @tgross?

Thanks!
